File: tasks-jvm/src/main/java/org/funfix/tasks/jvm/UncaughtExceptionHandler.java
package org.funfix.tasks.jvm;

/**
 * Utilities for handling uncaught exceptions.
 */
public final class UncaughtExceptionHandler {
    public static void rethrowIfFatal(final Throwable e) {
        if (e instanceof VirtualMachineError error) {
            throw error;
        }
        if (e instanceof ThreadDeath death) {
            throw death;
        }
        if (e instanceof LinkageError linkage) {
            throw linkage;
        }
    }

    public static void logOrRethrow(final Throwable e) {
        rethrowIfFatal(e);
        final var thread = Thread.currentThread();
        var logger = thread.getUncaughtExceptionHandler();
        if (logger == null) {
            logger = Thread.getDefaultUncaughtExceptionHandler();
        }
        if (logger == null) {
            logger = (t, e1) -> e1.printStackTrace(System.err);
        }
        logger.uncaughtException(thread, e);
    }
}

File: tasks/src/commonMain/kotlin/org/funfix/tasks/UncaughtExceptionHandler.kt
@file:Suppress("EXPECT_ACTUAL_CLASSIFIERS_ARE_IN_BETA_WARNING")

package org.funfix.tasks

import org.funfix.tasks.support.NonBlocking

@NonBlocking
public expect object UncaughtExceptionHandler {
    public fun rethrowIfFatal(e: Throwable)
    public fun logOrRethrow(e: Throwable)
}
